src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java

```diff
@@ -431,13 +431,18 @@ public void createDirectory(Path obj, FileAttribute<?>... attrs)
             dfd2 = dup(dfd1);
             dp = fdopendir(dfd1);
         } catch (UnixException x) {
-            if (dfd1 != -1)
-                UnixNativeDispatcher.close(dfd1);
-            if (dfd2 != -1)
-                UnixNativeDispatcher.close(dfd2);
-            if (x.errno() == UnixConstants.ENOTDIR)
-                throw new NotDirectoryException(dir.getPathForExceptionMessage());
-            x.rethrowAsIOException(dir);
+            IOException ioe = x.errno() == UnixConstants.ENOTDIR ?
+                new NotDirectoryException(dir.getPathForExceptionMessage()) :
+                x.asIOException(dir);
+            if (dfd1 != -1) {
+                ioe = UnixNativeDispatcher.close(dfd1, ioe,
+                    y -> y.asIOException(dir));
+            }
+            if (dfd2 != -1) {
+                ioe = UnixNativeDispatcher.close(dfd2, ioe,
+                    z -> z.asIOException(dir));
```

Comment:
It's not clear to me that the closing of dfd1 and dfd2 are interesting here, maybe they should be e -> null too.